### PR TITLE
new: `preserve_adb_keys` in PyG to ArangoDB

### PR DESCRIPTION
**Pull request by aMahanna**
Warrants a `1.1.0` release

<a href="https://colab.research.google.com/drive/16irAG6uQcWFXi-dfzA6eZUmeRghXjwy4?usp=sharing" target="_parent">Open In Colab</a>

Note: For notebook testing purposes:
* Use `!pip install git+https://github.com/arangoml/pyg-adapter.git@feature/adbpyg-map` instead of `!pip install adbpyg-adapter==1.1.0`
* Use `!git clone -b feature/adbpyg-map --single-branch https://github.com/arangoml/pyg-adapter.git` instead of `!git clone -b 1.1.0 --single-branch https://github.com/arangoml/pyg-adapter.git`

### What's new
Introduces the `preserve_adb_keys` parameter in ArangoDB to PyG. When set to `True`, will store the ArangoDB vertex & edge `_key` values in the (newly created) PyG graph. When calling PyG to ArangoDB, PyG graphs that have a `_key` or `_id` property will have their vertex & edge IDs persisted. See below for an example

#### Additional Changes
1. Feature: New supported metagraph value type for ArangoDB to PyG & PyG to ArangoDB: `Set[str]`
2. Bug: In PyG to ArangoDB, when `explicit_metagraph=True`, omitted PyG node & edge properties were still being brought over into ArangoDB (this has been fixed)
3. encapsulate some PyG to ArangoDB & ArangoDB to PyG logic in new helper methods: `__set_adb_data()`,`__set_pyg_data()`
4. code cleanup
5. docstring updates
6. test assertion logic refactor (the existing validation logic was out of date, as we were not accounting for the various supported values in our ArangoDB & PyG Metagraphs)
7. new: `pytest_exception_interact` in `conftest.py` for post-failure ArangoDB cleanup
8. new test cases
9. new `1.1.0` notebook

#### Demo (`preserve_adb_keys`)

```py
# With `preserve_adb_keys=True`, the adapter will preserve the ArangoDB vertex & edge _key values into the (newly created) PyG graph.
# Users can then re-import their PyG graph into ArangoDB using the same _key values 
pyg_g = adbpyg_adapter.arangodb_graph_to_pyg("imdb", preserve_adb_keys=True)

# pyg_g["Movies"]["_key"] --> ["1", "2", ..., "1682"]
# pyg_g["Users"]["_key"] --> ["1", "2", ..., "943"]
# pyg_g[("Users", "Ratings", "Movies")]["_key"] --> ["2732620466", ..., "2730643624"]

# Let's add a new PyG User Node by updating the _key property
pyg_g["Users"]["_key"].append("new-user-here-944")

# Note: Prior to the re-import, we must manually set the number of nodes in the PyG graph, since the `arangodb_graph_to_pyg` API creates featureless node data
pyg_g["Movies"].num_nodes = len(pyg_g["Movies"]["_key"]) # 1682
pyg_g["Users"].num_nodes = len(pyg_g["Users"]["_key"]) # 944 (prev. 943)

# Re-import PyG graph into ArangoDB
adbpyg_adapter.pyg_to_arangodb("imdb", pyg_g, on_duplicate="update")
```

#### Demo (`Set[str]` as a metagraph value)

```py
metagraph = {
    "nodeTypes": {
        "v0": {"x", "y"}, # equivalent to {"x": "x", "y": "y"}
    },
    "edgeTypes": {
        ("v0", "e0", "v0"): {"edge_attr"},
    },
}

adbpyg_adapter.pyg_to_arangodb("FakeData", data, metagraph, explicit_metagraph=False)

metagraph = {
    "vertexCollections": {
        "v0": {"x", "y"}, # equivalent to {"x": "x", "y": "y"}
    },
    "edgeCollections": {
        "e0": {"edge_attr"},
    },
}

pyg_g = adbpyg_adapter.arangodb_to_pyg("FakeData", metagraph)
```

### CODE REVIEW

File: adbpyg_adapter/adapter.py [hunk 1]
```diff
@@ -4,9 +4,14 @@
 from collections import defaultdict
 from typing import Any, DefaultDict, Dict, List, Set, Union
 
+try:
+    # https://github.com/arangoml/pyg-adapter/issues/4
+    from cudf import DataFrame
+except ModuleNotFoundError:
+    from pandas import DataFrame
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py]: ^ #4

File: adbpyg_adapter/adapter.py [hunk 2]
```diff
@@ -326,6 +345,18 @@ def pyg_to_arangodb(
...
+            PyG Node & Edge IDs back into ArangoDB Vertex & Edge IDs. Assumes that
+            the user has a valid ADB Map for **name**.
+            An ADB Map can be built by running ArangoDB to PyG operation for
+            the same **name**, or by creating the map manually. Defaults to False.
+
+            .. code-block:: python
+            adbpyg_adapter.pyg_to_arangodb(
+                name, pyg_g, preserve_adb_keys=True, on_duplicate="update"
+            )
+        :type preserve_adb_keys: bool
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py:359]: ^

File: adbpyg_adapter/adapter.py [hunk 3]
```diff
@@ -380,9 +412,31 @@ def v2_x_to_pandas_dataframe(t: Tensor):
...
         is_homogeneous = type(pyg_g) is Data
 
+        pyg_map = self.pyg_map[name]
+        if preserve_adb_keys:
+            if is_homogeneous:
+                msg = """
+                    **preserve_adb_keys** does not yet support
+                    homogeneous graphs (i.e type(pyg_g) is Data).
+                """
+                raise ValueError(msg)
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py:422]: ^

File: adbpyg_adapter/adapter.py [hunk 4]
```diff
@@ -415,13 +469,31 @@ def v2_x_to_pandas_dataframe(t: Tensor):
...
+            if preserve_adb_keys:
+                num_node_keys = len(pyg_map[n_type])
+
+                if num_nodes != num_node_keys:
+                    msg = f"""
+                        {num_nodes} does not match
+                        number of node keys in pyg_map
+                        ({num_node_keys}) for {n_type}
+                    """
+                    raise ValueError(msg)
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py:484]: ^

File: adbpyg_adapter/adapter.py [hunk 5]
```diff
@@ -636,25 +712,63 @@ def __build_tensor_from_dataframe(
 
         raise ADBMetagraphError(f"Invalid {meta_val} type")  # pragma: no cover
 
+    def __finish_adb_dataframe(
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py:715]: (this has nothing to do with `preserve_adb_keys`, just sliding this change in)

File: tests/test_adapter.py [hunk 6]
```diff
@@ -681,39 +731,39 @@ def assert_arangodb_data(
...
 
-        # TODO: Remove str restriction
-        has_node_feature_matrix = "x" in node_data and type(meta.get("x", "x")) is str
-        has_node_target_label = (
-            num_nodes == len(node_data.get("y", [])) and type(meta.get("y", "y")) is str
-        )
+        # # TODO: Remove str restriction
+        # has_node_feature_matrix = "x" in node_data and type(meta.get("x", "x")) is str
+        # has_node_target_label = (
+        #     num_nodes == len(node_data.get("y", [])) and type(meta.get("y", "y")) is str
```
Comment (aMahanna) [inline tests/test_adapter.py:737]: todo: re-enable feature/label in for `assert_arangodb_data` (this components needs to be refactored, as it is now out of date due to the new supported values in the adapter `metagraph`)

File: tests/test_adapter.py [hunk 7]
```diff
@@ -1,8 +1,10 @@
+# flake8: noqa
```
Comment (aMahanna) [inline tests/test_adapter.py:1]: todo: remove when https://github.com/arangoml/pyg-adapter/pull/11#discussion_r936215059 is handled

File: adbpyg_adapter/adapter.py (hunk 1 above)
Comment (aMahanna) [reply adbpyg_adapter/adapter.py]: reverted in [847ebae](https://github.com/arangoml/pyg-adapter/pull/11/commits/847ebae9fd904d5b7d0924a23929ea8b193d6557) due to https://github.com/arangoml/pyg-adapter/runs/7644978211?check_suite_focus=true (will need to further investigate)

File: adbpyg_adapter/adapter.py [hunk 8]
```diff
@@ -88,6 +91,22 @@ def arangodb_to_pyg(
...
+            adbpyg_adapter.pyg_to_arangodb(
+                graph_name, pyg_graph, ..., on_duplicate="update"
+            )
+
+            NOTE: If your ArangoDB graph is Homogeneous, the ArangoDB keys will
+            be preserved under `_v_key` & `_e_key` in your PyG graph. If your
+            ArangoDB graph is Heterogeneous, the ArangoDB keys will be preserved
+            under `_key` in your PyG graph.
+
+        :type preserve_adb_keys: bool
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py]: highlighting this to see if we can provide a better explanation

File: adbpyg_adapter/adapter.py [hunk 9]
```diff
@@ -636,33 +726,95 @@ def __build_tensor_from_dataframe(
...
+        df: DataFrame,
+        meta: Dict[Any, PyGMetagraphValues],
+        pyg_data: Union[Data, NodeStorage, EdgeStorage],
+        explicit_metagraph: bool,
+    ) -> DataFrame:
+        """A helper method to build the ArangoDB Dataframe for the given
+        collection. Is responsible for creating "sub-DataFrames" from PyG tensors
+        or lists, and appending them to the main dataframe **df**. If the data
+        does not adhere to the supported types, or is not of specific length,
+        then it is silently skipped.
```
Comment (aMahanna) [inline adbpyg_adapter/adapter.py:758]: will want to come back to this "silent" skip, leaving this as a note for now

**Review by joerg84 — APPROVED** [review]
I believe the entry Changelog is missing

**Comment by aMahanna** [comment]
> I believe the entry Changelog is missing

@joerg84 

No need, as the [release.yml](https://github.com/arangoml/pyg-adapter/blob/master/.github/workflows/release.yml#L90-L140) action will auto-generate a Changelog entry using commit messages upon triggering a new release (in this case it would be `1.1.0`). 

Example: https://github.com/arangoml/pyg-adapter/pull/9

**Pull request merged**